pkg/controller/elasticsearch/driver/upgrade_predicates.go

```diff
@@ -10,13 +10,98 @@ import (
+func getNodeSettings(
+	version version.Version,
+	resourcesList nodespec.ResourcesList,
+	pod corev1.Pod,
+) (esv1.ElasticsearchSettings, error) {
+	// Get the expected configuration
+	statefulSetName, _, err := sset.StatefulSetName(pod.Name)
+	if err != nil {
+		return esv1.ElasticsearchSettings{}, err
+	}
+	resources, err := resourcesList.ForStatefulSet(statefulSetName)
+	if err != nil {
+		return esv1.ElasticsearchSettings{}, err
+	}
+	nodeCfg, err := resources.Config.Unpack(version)
+	if err != nil {
+		return esv1.ElasticsearchSettings{}, err
+	}
+	return nodeCfg, nil
+}
+
+// getNodesSettings returns the node settings for a given list of Pods.
+func getNodesSettings(
+	version version.Version,
+	resourcesList nodespec.ResourcesList,
+	pods ...corev1.Pod,
+) ([]esv1.ElasticsearchSettings, error) {
+	rolesList := make([]esv1.ElasticsearchSettings, len(pods))
+	for i := range pods {
+		roles, err := getNodeSettings(version, resourcesList, pods[i])
+		if err != nil {
+			return nil, err
+		}
+		rolesList[i] = roles
+	}
+	return rolesList, nil
+}
+
+// hasDependencyInOthers returns true if, for a given node, at least one other node in a slice can be considered as a
+// strong dependency and must be upgraded first. A strong dependency is a unidirectional dependency, if a circular
+// dependency exists between two nodes the dependency is not considered as a strong one.
+func hasDependencyInOthers(node esv1.ElasticsearchSettings, others []esv1.ElasticsearchSettings) bool {
+	if !node.Node.CanContainData() {
+		// node has no tier which requires upgrade prioritization.
+		return false
+	}
+	for _, other := range others {
+		if !other.Node.CanContainData() {
+			// this other node has no tier which requires upgrade prioritization.
+			continue
+		}
+		if dependsOn(node.Node, other.Node) && !dependsOn(other.Node, node.Node) {
+			// candidate has this other node as a strict dependency
+			return true
+		}
+	}
+	// no dependency or roles are overlapping, we still allow the upgrade
+	return false
+}
+
+func dependsOn(candidate, other *esv1.Node) bool {
+	switch {
+	case !candidate.HasRole(esv1.MasterRole) && other.HasRole(esv1.MasterRole):
+		// other might be a dependency, but it is also a master node. We don't want to enter a deadlock where other is
+		// the last master node, while the candidate is not and must be upgraded first.
+		return false
+	case candidate.HasRole(esv1.DataHotRole):
+		// hot tier must be upgraded after warm, cold and frozen
+		return other.HasRole(esv1.DataWarmRole) || other.HasRole(esv1.DataColdRole) || other.HasRole(esv1.DataFrozenRole)
+	case candidate.HasRole(esv1.DataWarmRole):
+		// warm tier must be upgraded after cold and frozen
+		return other.HasRole(esv1.DataColdRole) || other.HasRole(esv1.DataFrozenRole)
+	case candidate.HasRole(esv1.DataColdRole):
+		// cold tier must be upgraded after frozen
+		return other.HasRole(esv1.DataFrozenRole)
+	}
+	// frozen and content have no dependency
+	return false
+}
```

Comment:
I was wondering wether the dependency logic could live closer to the domain model in the `esv1` package but no strong preference.